app/modules/softuart.c

```diff
@@ -0,0 +1,398 @@
+{
+    if ((gpio_id > 5 && gpio_id < 12) || gpio_id > 15)
+    {
+        return 0;
+    }
+    return 1;
+}
+
+void softuart_set_rx(uint8_t gpio_id)
+{
+    softuart.pin_rx = gpio_id;
+}
+
+void softuart_set_tx(uint8_t gpio_id)
+{
+    softuart.pin_tx = gpio_id;
+}
+
+//TODO ICACHE_RAM_ATTR
+static uint32_t softuart_intr_handler(uint32_t ret_gpio_status)
+{
+    Softuart* s = &softuart;
+    uint8_t level;
+    uint32_t mask = 1 << s->pin_rx;
+    uint32_t gpio_status = GPIO_REG_READ(GPIO_STATUS_ADDRESS);
+
+    // clear gpio status. Say ESP8266EX SDK Programming Guide in  5.1.6. GPIO interrupt handler
+    GPIO_REG_WRITE(GPIO_STATUS_W1TC_ADDRESS, gpio_status & mask);
+
+    if(gpio_status & mask) {
+        gpio_pin_intr_state_set(GPIO_ID_PIN(s->pin_rx), GPIO_PIN_INTR_DISABLE);
+
+        level = GPIO_INPUT_GET(GPIO_ID_PIN(s->pin_rx));
+        if(!level) {
+            //pin is low
+            //therefore we have a start bit
+
+            //wait till start bit is half over so we can sample the next one in the center
+            os_delay_us(s->bit_time/2);
+
+            //now sample bits
+            unsigned i;
+            unsigned d = 0;
+            unsigned start_time = 0x7FFFFFFF & system_get_time();
+
+            for(i = 0; i < 8; i ++ )
+            {
+                while ((0x7FFFFFFF & system_get_time()) < (start_time + (s->bit_time*(i+1))))
+                {
+                    //If system timer overflow, escape from while loop
+                    if ((0x7FFFFFFF & system_get_time()) < start_time){break;}
+                }
+                d >>= 1;
+
+                //read bit
+                if(GPIO_INPUT_GET(GPIO_ID_PIN(s->pin_rx))) {
+                    //if high, set msb of 8bit to 1
+                    d |= 0x80;
+                }
+            }
+
+            //store byte in buffer
+            uint8 next = s->buffer.length + 1 % SOFTUART_MAX_RX_BUFF;
+            if (next != 0)
+            {
+                s->buffer.receive_buffer[s->buffer.length] = d; // save new byte
+                s->buffer.length = next;
+
+                NODE_DBG("SOFTUART char received - new buffer length %d\r\n", s->buffer.length);
+                if( ((s->need_len != 0) && (s->buffer.length >= s->need_len)) || \
+                    ((s->end_char>=0) && ((unsigned char)d==(unsigned char)s->end_char)))
+                {
+                    softuart_on_data_cb(s->buffer.receive_buffer, s->buffer.length);
+                    s->buffer.length = 0;
+                }
+
+            }
+            else
+            {
+                NODE_DBG("SOFTUART buffer overflow\r\n");
+            }
+
+            //wait for stop bit
+            os_delay_us(s->bit_time);
+
+            //done
+        }
+
+        //clear interrupt
+        GPIO_REG_WRITE(GPIO_STATUS_W1TC_ADDRESS, gpio_status);
+
+        // Reactivate interrupts for GPIO0
+        gpio_pin_intr_state_set(GPIO_ID_PIN(s->pin_rx), GPIO_PIN_INTR_ANYEDGE);
```

Comment:
Why triggering on ANY edge. If inverted mode not supported, then triggering is on FALLING edge only.